### PR TITLE
Support multiple artifact storage backends

### PR DESCRIPTION
**Pull request by dsuhinin**

### CODE REVIEW

**Comment by dsuhinin** [comment]
> Just scanned it very quickly and it looks to me like we will create a new storage provider on every request to list artifacts. For the S3 provider, I believe this would mean recreating a client (and thus connections) every time.
> 
> I think we should instead register each storage provider with a URL scheme and instantiate them (lazily?) in the "factory" so they can be returned based on the scheme of the requested URI. So a proxy instead of a factory I guess.

emmm, I understand what you mean. in general even now there is no heavy lifting but let me think.

File: pkg/api/mlflow/config/config_test.go [hunk 1]
```diff
@@ -17,19 +17,19 @@ func TestServiceConfig_Validate_Ok(t *testing.T) {
 		{
 			name: "ArtifactRootHasS3Prefix",
 			providedConfig: &ServiceConfig{
-				ArtifactRoot: "s3://bucket_name",
+				DefaultArtifactRoot: "s3://bucket_name",
```
Comment (jgiannuzzi) [inline pkg/api/mlflow/config/config_test.go:20]: Please update all the test case names as well

File: pkg/cmd/import.go [hunk 2]
```diff
@@ -74,7 +74,7 @@ func init() {
 
 	ImportCmd.Flags().StringP("input-database-uri", "i", "", "Input Database URI (eg., sqlite://fasttrackml.db)")
 	ImportCmd.Flags().StringP("output-database-uri", "o", "", "Output Database URI (eg., postgres://user:psw@postgres:5432)")
-	ImportCmd.Flags().StringP("artifact-root", "a", "./artifacts", "Artifact Root")
+	ImportCmd.Flags().StringP("default-artifact-root", "a", "./artifacts", "Artifact Root")
```
Comment (jgiannuzzi) [inline pkg/cmd/import.go:77]: please also update the lines above, most importantly line 44!

File: pkg/api/mlflow/config/config.go [hunk 3]
```diff
@@ -34,8 +34,8 @@ func NewServiceConfig() *ServiceConfig {
 		ListenAddress:         viper.GetString("listen-address"),
 		AuthUsername:          viper.GetString("auth-username"),
 		AuthPassword:          viper.GetString("auth-password"),
-		ArtifactRoot:          viper.GetString("artifact-root"),
 		S3EndpointURI:         viper.GetString("s3-endpoint-uri"),
+		DefaultArtifactRoot:   viper.GetString("default-artifact-root"),
```
Comment (jgiannuzzi) [inline pkg/api/mlflow/config/config.go:38]: if you move it down, maybe sort the list?

File: pkg/cmd/server.go [hunk 4]
```diff
@@ -208,7 +207,7 @@ func init() {
 	RootCmd.AddCommand(ServerCmd)
 
 	ServerCmd.Flags().StringP("listen-address", "a", "localhost:5000", "Address (host:post) to listen to")
-	ServerCmd.Flags().String("artifact-root", "./artifacts", "Artifact root")
+	ServerCmd.Flags().String("default-artifact-root", "./artifacts", "Artifact root")
```
Comment (jgiannuzzi) [inline pkg/cmd/server.go:210]:
```suggestion
	ServerCmd.Flags().String("default-artifact-root", "./artifacts", "Default artifact root")
```

**Review by jgiannuzzi — CHANGES_REQUESTED** [review]
Please also update `db_factory.go`, `db_instance.go`, and `import_test.go`

**Review by jgiannuzzi — COMMENTED** [review]
Please also update `db_factory.go`, `db_instance.go`, and `import_test.go`

File: pkg/api/mlflow/config/config_test.go (hunk 1 above)
Comment (dsuhinin) [reply pkg/api/mlflow/config/config_test.go:20]: ahh, shit. yes.

File: pkg/api/mlflow/config/config.go (hunk 3 above)
Comment (dsuhinin) [reply pkg/api/mlflow/config/config.go:38]: I though about that too =) but since it is a bit related to `s3-endpoint`, decided to keep it here.

Comment (dsuhinin) [reply pkg/api/mlflow/config/config.go:38]: but ok, let me move it.

File: pkg/api/mlflow/config/config.go [hunk 5]
```diff
@@ -35,11 +35,11 @@ func NewServiceConfig() *ServiceConfig {
 		AuthUsername:          viper.GetString("auth-username"),
 		AuthPassword:          viper.GetString("auth-password"),
 		S3EndpointURI:         viper.GetString("s3-endpoint-uri"),
-		DefaultArtifactRoot:   viper.GetString("default-artifact-root"),
 		DatabaseURI:           viper.GetString("database-uri"),
 		DatabaseReset:         viper.GetBool("database-reset"),
 		DatabasePoolMax:       viper.GetInt("database-pool-max"),
 		DatabaseMigrate:       viper.GetBool("database-migrate"),
+		DefaultArtifactRoot:   viper.GetString("default-artifact-root"),
```
Comment (jgiannuzzi) [inline pkg/api/mlflow/config/config.go:42]: This position makes even less sense 😂

Comment (dsuhinin) [reply pkg/api/mlflow/config/config.go:42]: correct.

Comment (dsuhinin) [reply pkg/api/mlflow/config/config.go:42]: lets keep it like before.

**Review by jgiannuzzi — APPROVED** [review]

**Pull request merged**